File: 86.partition-list.java
/*
 * @lc app=leetcode id=86 lang=java
 *
 * [86] Partition List
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode partition(ListNode head, int x) {

        ListNode leftHead = new ListNode(0, null),
                 left = leftHead, 
                 rightHead = new ListNode(0, null),
                 right = rightHead;


        while (head != null) {
            if (head.val < x) {
                left.next = head;
                left = left.next;
            } else {
                right.next = head;
                right = right.next;
            }

            head = head.next;
        }

        right.next = null;
        left.next = rightHead.next;

        return leftHead.next;
        
    }
}
// @lc code=end

File: 86.partition-list.py
#
# @lc app=leetcode id=86 lang=python3
#
# [86] Partition List
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def partition(self, head: Optional[ListNode], x: int) -> Optional[ListNode]:
        leftHead = ListNode(0, None)
        left = leftHead
        rightHead = ListNode(0, None)
        right = rightHead

        while head:
            if head.val < x:
                left.next = head
                left = left.next
            else:
                right.next = head
                right = right.next
            
            head = head.next
        
        right.next = None
        left.next = rightHead.next

        return leftHead.next

        
# @lc code=end

